[pulsar-admin] Add remove-clusters command for namespace

yuruguo opened the pull request:
From [dev-email](https://lists.apache.org/thread/z531k47nlzy3gvbdkthh2n2wblwgxzfs)
Fixes #12822 

### Motivation
CLI `bin/pulsar-admin` supports `set-clusters` and `get-clusters` command, but lacks corresponding `remove-clusters` commands,  the purpose of this PR is to add this command.

### Documentation
Automatically generate doc through code
- [x] `doc`

yuruguo commented:
@codelipenghui @congbobo184 @hangc0276 PTAL, thx!

yuruguo commented:
/pulsarbot run-failure-checks

yuruguo commented:
/pulsarbot run-failure-checks

codelipenghui commented:
@yuruguo Can you add a test to make sure all the clusters of a namespace can be removed and the topic can still work for publishing and consumption?

yuruguo commented:
/pulsarbot run-failure-checks

yuruguo commented:
@merlimat @eolivelli @aahmed-se @codelipenghui @315157973 @hangc0276 @gaoran10 PTAL

yuruguo commented:
@merlimat @eolivelli @aahmed-se @k2la @codelipenghui @congbobo184 @hangc0276 @gaoran10 PTAL

yuruguo commented:
@eolivelli @codelipenghui PTAL

yuruguo commented:
/pulsarbot run-failure-checks

yuruguo commented:
/pulsarbot run-failure-checks

yuruguo commented:
/pulsarbot run-failure-checks

yuruguo commented:
/pulsarbot run-failure-checks

yuruguo commented:
/pulsarbot run-failure-checks

yuruguo commented:
/pulsarbot run-failure-checks

yuruguo commented:
/pulsarbot run-failure-checks

yuruguo commented:
@eolivelli @aahmed-se @codelipenghui @hangc0276 PTAL

yuruguo commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/Namespaces.java:
```diff
@@ -271,6 +272,18 @@ public void setNamespaceReplicationClusters(@PathParam("tenant") String tenant,
+    public void removeNamespaceReplicationClusters(@PathParam("tenant") String tenant,
+                                                    @PathParam("namespace") String namespace) {
+        validateNamespaceName(tenant, namespace);
+        internalSetNamespaceReplicationClusters(Lists.newArrayList());
```
I understand that `replication_clusters` is not `null`, because it has been assigned when the namespace is created for  v2 namespace, as follows:
https://github.com/apache/pulsar/blob/3e6fedf9b69758c13d92c6ff75a7bd779038543a/pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java#L131-L134
https://github.com/apache/pulsar/blob/3e6fedf9b69758c13d92c6ff75a7bd779038543a/pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java#L2025-L2029
and Its default value is `HashSet` for v1 namespace, as below:
https://github.com/apache/pulsar/blob/3e6fedf9b69758c13d92c6ff75a7bd779038543a/pulsar-client-admin-api/src/main/java/org/apache/pulsar/common/policies/data/Policies.java#L38
So I am a little puzzled about the introduction of `state`

eolivelli commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v1/Namespaces.java:
```diff
@@ -342,6 +342,19 @@ public void setNamespaceReplicationClusters(@PathParam("property") String proper
 
+    @DELETE
+    @Path("/{property}/{cluster}/{namespace}/replication")
+    @ApiOperation(value = "Set the replication clusters for namespace")
```
Please fix the help

yuruguo replied:
done

eolivelli commented on pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdNamespaces.java:
```diff
@@ -325,6 +325,18 @@ void run() throws PulsarAdminException {
+    @Parameters(commandDescription = "Remove replication clusters for a namespace")
+    private class RemoveReplicationClusters extends CliCommand {
+        @Parameter(description = "tenant/namespace", required = true)
+        private java.util.List<String> params;
```
I cannot understand this parameter

yuruguo replied:
The `params` represents the `tenant/namespace`, although it is a `List` type but actually only allows `single` value, as below:
https://github.com/apache/pulsar/blob/10818e8e39b9f039793b8534da0252f2f4c91a61/pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CliCommand.java#L47-L48
https://github.com/apache/pulsar/blob/10818e8e39b9f039793b8534da0252f2f4c91a61/pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CliCommand.java#L122-L128

tisonkun replied:
@yuruguo I suggest you provide some command-line examples here to describe where is this param place :)

yuruguo commented:
/pulsarbot run-failure-checks

yuruguo commented:
/pulsarbot run-failure-checks

yuruguo commented:
@eolivelli @codelipenghui @315157973 please take a look again, thx!

eolivelli commented:
@yuruguo we are adding more and more new REST endpoints in order to support this set of commands for pulsar-admin.

What about creating a list of the features that you want to implement and discuss it with the community.

The risk is to add more and more APIs without control and discussions in the community.

yuruguo commented:
> @yuruguo we are adding more and more new REST endpoints in order to support this set of commands for pulsar-admin.
> 
> What about creating a list of the features that you want to implement and discuss it with the community.
> 
> The risk is to add more and more APIs without control and discussions in the community.

Thank you for your reminder and guidance, I understand what you mean.
In fact, the `remove` operation is a dangerous operation. For the new feature, I will first send an email to discuss in @dev, and then start the development work after reaching an agreement.

yuruguo commented:
> @yuruguo we are adding more and more new REST endpoints in order to support this set of commands for pulsar-admin.
> 
> What about creating a list of the features that you want to implement and discuss it with the community.
> 
> The risk is to add more and more APIs without control and discussions in the community.

Thanks for your reminder and guidance, I understand what you mean.
In fact, the `remove` operation is a dangerous operation. For the feature, I will first send an email to discuss in @dev, and then start the development work after reaching an agreement.

yuruguo commented:
> @yuruguo we are adding more and more new REST endpoints in order to support this set of commands for pulsar-admin.
> 
> What about creating a list of the features that you want to implement and discuss it with the community.
> 
> The risk is to add more and more APIs without control and discussions in the community.

Thanks for your remind and guidance, I understand what you mean. What I want to say is,

1. Can you provide more information, such as related issue / pr / dev-email for `we are adding more and more new REST endpoints in order to support this set of commands for pulsar-admin`, because I want to see its implementation, thx!
2. Second, for this feature, I will create a PR and send email to @dev to discuss with community, then decide the follow-up work based on the results of the discussion. And for the new features that may be added later, I will continue to use this method :)

yuruguo commented:
@codelipenghui @hangc0276 ptal

yuruguo commented:
@eolivelli @codelipenghui @315157973 @hangc0276 ptal

yuruguo commented:
@codelipenghui @hangc0276 @gaoran10 PTAL, thx!

yuruguo commented:
@codelipenghui @hangc0276 @BewareMyPower PTAL, thx!

yuruguo commented:
@eolivelli @codelipenghui @315157973 can you help review this pr again?

yuruguo commented:
@eolivelli @codelipenghui @315157973 can you help review this pr again?

yuruguo commented:
@eolivelli @codelipenghui @315157973 can you help review this pr again?

yuruguo commented:
@eolivelli @codelipenghui @315157973 can you help review this pr again?

yuruguo commented:
@eolivelli @codelipenghui @315157973 can you help review this pr again? I have addressed all comments :)

yuruguo commented:
@eolivelli @codelipenghui @315157973 can you help review this pr again? I have addressed all comments :)

yuruguo commented:
@eolivelli @codelipenghui @315157973 can you help review this pr again? I have addressed all comments :)

yuruguo commented:
@eolivelli @codelipenghui @315157973 can you help review this pr again? I have addressed all comments :)

yuruguo commented:
@eolivelli @codelipenghui @315157973 Would you help review this pr again? I have addressed all comments :)

yuruguo commented:
@eolivelli @codelipenghui @315157973 Would you help review this pr again? I have addressed all comments :)

yuruguo commented:
@eolivelli @codelipenghui @315157973 Sorry, Could you review this pr again? I have addressed all comments :)

315157973 commented:
> @eolivelli @codelipenghui @315157973 Sorry, Could you review this pr again? I have addressed all comments :)

Please follow eolivelli‘s comment. It may be a risk is to add more and more APIs without control and should discuss in the community.

yuruguo commented:
> > @eolivelli @codelipenghui @315157973 Sorry, Could you review this pr again? I have addressed all comments :)
> 
> Please follow eolivelli‘s comment. It may be a risk is to add more and more APIs without control and should discuss in the community.

Thanks for your reply, I have created a related [issue](https://github.com/apache/pulsar/issues/12822) and sent an email to the [dev mailing list](https://lists.apache.org/thread/z531k47nlzy3gvbdkthh2n2wblwgxzfs) to discuss it.
Maybe there are other tasks I am happy to do :)

yuruguo commented:
/pulsarbot run-failure-checks

poorbarcode commented:
Since we will start the RC version of `3.0.0` on `2023-04-11`, I will change the label/milestone of PR who have not been merged.
- The PR of type `feature` is deferred to `3.1.0`
- The PR of type `fix` is deferred to `3.0.1`

So drag this PR to `3.1.0`
